Boost: Handle cache gc in a scheduled event

haqadn opened the pull request:
Fixes HOG-26

## Proposed changes:
* Handle cache cleanup due to page output change in a scheduled event
* **Changed garbage collection disable value**. Previously you'd `define( 'JETPACK_BOOST_CACHE_DURATION', 0 )` to disable garbage collection. Now, you need to `define( 'JETPACK_BOOST_CACHE_DURATION', -1 )`

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
No

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
* Activate Boost Cache
* Toggle different modules and see garbage collection is triggering from scheduled jobs
  * Enable cache debug logging to see if garbage collection is happeneing
  * To see if there are any scheduled jobs consider disabling cron system to prevent the scheduled events from executing `define('DISABLE_WP_CRON', true) ;`

donnchawp committed:
We need to remove directories when rebuilding too

GC can delete rebuild files, leaving empty directories. Removing this
check makes sure those directories are deleted.

donnchawp committed:
Return directoroes in iterator too.

donnchawp committed:
changelog

donnchawp committed:
Add tests for Filesyste_Utils.php

donnchawp committed:
Fix the type of the $type parameter as static analysis checks failed

donnchawp committed:
Checks seem to be stalled. Kick them?

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/fix-deleting-directories

donnchawp committed:
Updating PHAN baseline

donnchawp committed:
Delete index.html files when deleting everything

If that file is left over, then the directory can't be deleted

haqadn committed:
Refactor walk_directory and garbage collection

haqadn committed:
Decouple file-storage from cache

haqadn committed:
Reuse delete and rebuild path actions

donnchawp committed:
Fix this typo and it's one file

donnchawp committed:
Delete the cache file for this user as they left a comment

We need to delete this file instead of rebuild as the cache file is only
used by that particular user because of their cookies.

donnchawp committed:
Add $parameters parameter to rebuild and delete page functions

This allows those functions to delete a single page visited by a
particular visitor with cookies.

donnchawp committed:
Apply the action to the root $path too.

haqadn committed:
Re-use some code

haqadn committed:
Remove unused method

haqadn committed:
Refactor the storage interface

haqadn committed:
Refactor rebuilding

haqadn committed:
Schedule a garbage collection when page output changes

haqadn committed:
Merge branch 'trunk' into update/schedule-gc

haqadn committed:
Fix comment approval check

haqadn committed:
Update changelog

LiamSarsfield commented:
A question, and a fix for Phan:
When the `activate` of the Garbage_Collection is called, should the schedule have the `JETPACK_BOOST_CACHE_DURATION` for an argument?
https://github.com/Automattic/jetpack/blob/0d5710ec04172ca4ff213b745c41107450b619f3/projects/plugins/boost/app/modules/optimizations/page-cache/class-garbage-collection.php#L36-L37
to
`wp_schedule_event( time(), 'hourly', self::ACTION, array( JETPACK_BOOST_CACHE_DURATION ) )`

Phan error can be fixed by changing
https://github.com/Automattic/jetpack/blob/0d5710ec04172ca4ff213b745c41107450b619f3/projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/storage/interface-storage.php#L18
to
`public function garbage_collect( int $cache_duration = JETPACK_BOOST_CACHE_DURATION );`

I can't suggest neither within GitHub unfortuntently, because are lines of code changed
LGTM otherwise.

LiamSarsfield reviewed (changes requested):
A question, and a fix for Phan:
When the `activate` of the Garbage_Collection is called, should the schedule have the `JETPACK_BOOST_CACHE_DURATION` for an argument?
https://github.com/Automattic/jetpack/blob/0d5710ec04172ca4ff213b745c41107450b619f3/projects/plugins/boost/app/modules/optimizations/page-cache/class-garbage-collection.php#L36-L37
to
`wp_schedule_event( time(), 'hourly', self::ACTION, array( JETPACK_BOOST_CACHE_DURATION ) )`

Phan error can be fixed by changing
https://github.com/Automattic/jetpack/blob/0d5710ec04172ca4ff213b745c41107450b619f3/projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/storage/interface-storage.php#L18
to
`public function garbage_collect( int $cache_duration = JETPACK_BOOST_CACHE_DURATION );`

I can't suggest neither within GitHub unfortuntently, because are lines of code changed
LGTM otherwise.

haqadn committed:
Fix single garbage collection

haqadn commented:
> A question, and a fix for Phan: When the `activate` of the Garbage_Collection is called, should the schedule have the `JETPACK_BOOST_CACHE_DURATION` for an argument?
> 
> https://github.com/Automattic/jetpack/blob/0d5710ec04172ca4ff213b745c41107450b619f3/projects/plugins/boost/app/modules/optimizations/page-cache/class-garbage-collection.php#L36-L37
> 
> 
> to
> `wp_schedule_event( time(), 'hourly', self::ACTION, array( JETPACK_BOOST_CACHE_DURATION ) )`
> Phan error can be fixed by changing
> 
> https://github.com/Automattic/jetpack/blob/0d5710ec04172ca4ff213b745c41107450b619f3/projects/plugins/boost/app/modules/optimizations/page-cache/pre-wordpress/storage/interface-storage.php#L18
> 
> 
> to
> `public function garbage_collect( int $cache_duration = JETPACK_BOOST_CACHE_DURATION );`
> I can't suggest neither within GitHub unfortuntently, because are lines of code changed LGTM otherwise.

I wanted to make sure "clear cache" request doesn't apply to files created after the request was accepted. After your comment I realized that it broke while I was fixing merge conflict. The intention also wasn't clear. So, I have added a comment and fix the functionality. We can't provide a value in `wp_schedule_event` in this case because we don't want to fix the garbage collection to happen only for files older than when it was initially scheduled. The confusion came because of bad variable names, I used. The value is supposed to be a timestamp, and not a ttl. Hopefully, the changes will now make more sense.

haqadn commented:
Seems like cache e2e test is failing :/

haqadn committed:
Finish cache to account for cache preloading

LiamSarsfield committed:
Merge branch 'trunk' into update/schedule-gc

LiamSarsfield reviewed (changes requested):
Sorry I'm still getting an error when toggling modules:
`Uncaught Error: Unknown named parameter $older_than`
After I (I as in AI) looked into it (A)I found the culprit:
Basically, the scheduled task (jetpack_boost_cache_garbage_collection) tries to run both the main garbage_collect function and the Logger::delete_old_logs function. It passes the older_than timestamp value to both of them.
The problem is, Logger::delete_old_logs wasn't set up to accept any arguments. So when WordPress tried to give it the timestamp, PHP got confused and threw the error inside the WP hook system. It's just a mismatch between what the function expected and what the cron action tried to pass it.

LiamSarsfield commented:
I've merged the E2E test fix into trunk separately within #43195 so other PRs are less flaky

kraftbj committed:
Merge branch 'trunk' into update/schedule-gc

matticbot committed:
Update dependency @mdn/browser-compat-data to v6 (#43199)

Co-authored-by: Renovate Bot <bot@renovateapp.com>

kraftbj committed:
Add support for sqlite in wordbless (#43197)

anomiex committed:
phpcs: Fix and enable `Jetpack.PHPUnit.Attributes` sniff (#43175)

Mostly this is the result of running phpcbf on files failing the sniff.

Things needing manual cleanup were:

* `projects/packages/image-cdn/tests/php/Image_CDN_Test.php`:
  Remove a broken `@covers` that had no space between the tag and class
  name.

* `projects/plugins/jetpack/tests/php/core-api/WPCOM_REST_API_V2_Field_Controller_Test.php`:
  Fix a `@group` that was accidentally
  "rest-api\n\nphpcs:disable Generic.Files.OneObjectStructurePerFile.MultipleFound".

* `projects/plugins/jetpack/tests/php/core-api/wpcom-fields/WPCOM_REST_API_V2_Attachment_VideoPress_Data_Test.php`:
* `projects/plugins/jetpack/tests/php/sync/Jetpack_Sync_Queue_TestBase.php`:
  Move some `@group`s from the file comment to class comment.

* `projects/plugins/jetpack/tests/php/Functions_OpenGraph_Test.php`:
* `projects/plugins/jetpack/tests/php/general/Jetpack_Client_Server_Test.php`:
  Fix where it got confused about whether a comment was a file or class
  comment.

* `projects/plugins/jetpack/tests/php/json-api/Jetpack_Json_Api_Plugins_Endpoints_Test.php`:
  Fix `@covers` for nonexistent class.

* `projects/plugins/jetpack/tests/php/modules/shortcodes/trait.http-request-cache.php`:
* `projects/plugins/jetpack/tests/php/trait-woo-tests.php`:
  Manually fix, since the sniff doesn't recognize the trait as a test
  class.

* Disable `PHPCompatibility.Attributes.NewAttributes.PHPUnitAttributeFound` in phpcompatibility-dev run.

`PHPCompatibility.Attributes.NewAttributes.Found` was already disabled.

anomiex committed:
phpunit: Enable reporting and failing on PHPUnit 11 deprecations (#43179)

Now that we've fixed all the PHPUnit 11 deprecations, adjust the configs
to enable `displayDetailsOnPhpunitDeprecations` and
`failOnPhpunitDeprecation`.

For plugins/crm, plugins/jetpack, and plugins/wpcomsh, there's one more
deprecation left that's also fixed in this PR: WP Core's obsolete
`WP_UnitTestCase_Base::checkRequirements()` function has a doc comment
that PHPUnit is mis-parsing to see a `@group` annotation that's not
actually intended. We can override it to stop PHPUnit from seeing that
doc comment.

matticbot committed:
Update Eslint packages (#43200)

* Update Eslint packages
* Ignore another package with a build dep trying to `npm install` something
* Remove no-longer-needed eslint-disable directives

---------

Co-authored-by: Renovate Bot <bot@renovateapp.com>
Co-authored-by: Brad Jorsch <brad.jorsch@automattic.com>

edanzer committed:
Forms: Change data structure for integrations endpoint (#43183)

johngodley committed:
Gravatar: add Gravatar profiles to comments form (#42458)

* Add initial Gravatar comments

* Add changelog

* Revert webpack changes

Problem fixed in hovercard package

* Add i18n

* Use latest hovercard library

* Fix flex on email field

On simple sites it seems to be different

* Fix changelog?

* Bump to 0.8.0 Quick Editor

Track where interaction is coming from

* Lazy load Gravatar profile

* CSS lint fix

fgiannar committed:
Jetpack Sync: Switch 'jetpack_package_version' to a callable (#43188)

* Jetpack Sync: Switch 'jetpack_package_version' to a callable

johngodley committed:
Gravatar comments: fix hovercard style (#43207)

* Gravatar: fix hovercard style

Ensure the hovercard style is loaded in the main CSS

* Add changelog

enejb committed:
Update/file field interactivity fixes (#43201)

coder-karen committed:
Image Compare block: Ensure compare line is not visible on top of a sticky menu (#43187)

coder-karen committed:
General: Bump minimum WordPress version to 6.7 (#43192)

josephscott committed:
API Update Post: Fix warning when `$input['type']` does not exist (#43180)

jeherve committed:
e2e Slack notifications: update file upload logic (#43196)

* e2e Slack notifications: update file upload logic

Slack will soon deprecate the files.upload API:
https://docs.slack.dev/changelog/2024-04-a-better-way-to-upload-files-is-here-to-stay/

We are encouraged to move to files.getUploadURLExternal and files.completeUploadExternal instead.
https://api.slack.com/methods/files.getUploadURLExternal
https://api.slack.com/methods/files.completeUploadExternal

* changelog

* Expand the tests to avoid coverage failure

* changelog: update wording.

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

* Update implementation to use @slack/web-api

See https://github.com/Automattic/jetpack/pull/43196#pullrequestreview-2784732826

* Additional error check

See https://github.com/Automattic/jetpack/pull/43196/files#r2054583550

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

* Bring filesUploadV2 back into postOrUpdateMessage

We don't need to extract things anymore, since we simplified the upload in bf38b5fdcfed50547db4919ad5075f012d00a72c

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

* Update test

* Remove extra handling of errors

---------

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

haqadn committed:
Fix extra parameter issue

haqadn committed:
Remove unused const

haqadn committed:
Merge remote-tracking branch 'origin/trunk' into update/schedule-gc

haqadn commented:
> Sorry I'm still getting an error when toggling modules: `Uncaught Error: Unknown named parameter $older_than` After I (I as in AI) looked into it (A)I found the culprit: Basically, the scheduled task (jetpack_boost_cache_garbage_collection) tries to run both the main garbage_collect function and the Logger::delete_old_logs function. It passes the older_than timestamp value to both of them. The problem is, Logger::delete_old_logs wasn't set up to accept any arguments. So when WordPress tried to give it the timestamp, PHP got confused and threw the error inside the WP hook system. It's just a mismatch between what the function expected and what the cron action tried to pass it.

Good catch! I've addressed the issue in [2b48083](https://github.com/Automattic/jetpack/pull/42954/commits/2b48083f50e0591afe54c63a0d9b4391f90bfd35)

LiamSarsfield committed:
Merge branch 'trunk' into update/schedule-gc

LiamSarsfield commented on projects/plugins/boost/app/modules/optimizations/page-cache/class-garbage-collection.php:
```diff
@@ -6,16 +6,42 @@
+		add_action( self::ACTION, array( self::class, 'garbage_collect' ) );
+
+		// Clear old log files when garbage collection is run. Do not pass the $older_than parameter to the method as it's not supported.
+		add_action( self::ACTION, array( Logger::class, 'delete_old_logs' ), 10, 0 );
```
Neat! I never saw this pattern before of purposely adding no arguments, makes sense.

LiamSarsfield reviewed (approved):
LGTM 🚀 Tested it locally, and on a JN site, and no error logs reported when toggling modules or running the GC cronjob manually.
